File: SQP/EQP_Solution.m
function [p, lambda] = EQP_Solution(G,A,d,b,theta)
%EQP_SOLUTION 此处显示有关此函数的摘要
%  这个函数负责解每个iterate里头的EQP，分析发现，其实每个EQP中不一样的东西
%  只有A与b中的内容物，然后直接用KKT矩阵也可以一步到位得到相应的解，也就是
%  返回了当前的lambda值和相应的步长
%  G 为 n X n 的objective function中的二次项的系数
%  A 为 b X n 的当前在active set中的constraints们的theta上的参数
%  d 为原objective function中的与参数相乘的常数项，n X 1 的列向量
%  b 为constraint中的常数组成的列向量，在等式的右边
size_A =size(A);

% construction of components of KKT-matrix
g = d + G * theta;
h = A * theta - b ;
K = [G  A';A zeros(size_A(1, 1), size_A(1, 1))];

result = pinv(K)* [ g ; h ];

p = result(1:size_A(1,2), 1);
lambda = result( size_A(1, 2) + 1 : end, 1);

end

File: python/EQP_Solution.py
# A has a dimension of (m,n) m row, n column.
# Each row contain parameters of constraints. There are totally m constraints.
# b has a dimension of (m,1) m row, 1 column
# G has a dimension of (n,n) n row, n column
# d has a dimension of (n,1) n row, 1 column
# theta has a dimension of (n,1) n row, 1 column

import numpy as np

def EQP_Solution (A, b, G, d, theta):
    g = np.dot(G, theta) + d
    print("current g is :{}".format(g))
    h = np.dot(A, theta) - b
    print("current h is :{}".format(h))

    # concatenate the matrix into a KKT-Matrix and [g;h] vector
    [row, col] = np.shape(A)
    K_temp1 = np.concatenate((G, np.transpose(A)), axis=1)
    zeros_matrix = np.zeros([row, row])
    K_temp2 = np.concatenate((A, zeros_matrix), axis=1)

    K = np.concatenate((K_temp1, K_temp2), axis=0)
    print("current K is : {}".format(K))
    g_h = np.concatenate((g, h), axis=0)

    # calculation the step length and lambda number
    K_inv = np.linalg.inv(K)
    result = np.dot(K_inv, g_h)

    # from result separate the step length and lambda value
    step_length = -result[0:col, 0].reshape(col, 1)
    lambda_star = result[col:, 0].reshape(row, 1)

    return step_length, lambda_star